Sources/NIOSSL/NIOSSLHandler.swift

```diff
@@ -200,20 +255,70 @@ public class NIOSSLHandler : ChannelInboundHandler, ChannelOutboundHandler, Remo
             self.bufferFlush()
             self.doUnbufferWrites(context: context)
         }
     }
-    
+
     public func close(context: ChannelHandlerContext, mode: CloseMode, promise: EventLoopPromise<Void>?) {
-        guard mode == .all else {
-            // TODO: Support also other modes ?
+        switch mode {
+        case .output:
+            self.closeOutput(context: context, promise: promise)
+        case .all:
+            self.closeAll(context: context, promise: promise)
+        default:
             promise?.fail(ChannelError.operationUnsupported)
-            return
         }
-        
+    }
+
+    private func closeOutput(context: ChannelHandlerContext, promise: EventLoopPromise<Void>?) {
+        switch state {
+        case .closingOutput:
+            // We're in the process of closing the output, so let's let that happen. However,
+            // we want to cascade the result of the first request into this new one.
+            if let promise = promise, let closeOutputPromise = self.closeOutputPromise {
+                closeOutputPromise.futureResult.cascade(to: promise)
+            } else if let promise = promise {
+                self.closeOutputPromise = promise
+            }
+        case .closing:
+            // We're in the process of TLS shutdown, which has a higher priority.
+            // Therefore we skip the output closing procedure and cascade the result
+            // of the TLS shutdown request to this new one.
+            if let promise = promise, let closePromise = self.closePromise {
+                closePromise.futureResult.cascade(to: promise)
+            } else if let promise = promise {
+                self.closePromise = promise
+            }
+        case .idle, .outputClosed, .closed, .unwrapping, .unwrapped:
+            // For idle, outputClosed, closed, unwrapping, and unwrapped connections we immediately pass this on to the next
+            // channel handler.
+            context.close(mode: .output, promise: promise)
+        case .active, .handshaking, .additionalVerification, .inputClosed:
+            // It may occur that we are still in the process of handshaking or additional verification. We'll let that happen.
+            // We flush all outstanding writes once the handshake step is complete and set our state to .outputClosed aftwerwards.
+            // This prevents any further writes to this channel.
+            if let promise = promise, let closeOutputPromise = self.closeOutputPromise {
+                closeOutputPromise.futureResult.cascade(to: promise)
+            } else if let promise = promise {
+                self.closeOutputPromise = promise
+                self.state = .closingOutput(self.scheduleTimedOutShutdown(context: context))
+                self.doShutdownStep(context: context)
+            }
+
+            self.flush(context: context)
+
+            self.closeOutputPromise?.futureResult.whenSuccess {
```

Comment:
This is incorrect: if we don't have a promise, this never fires.

This makes me thing that maybe `outputClosing` isn't necessary at all. It's a sort of immediate transition: we send our CLOSE_NOTIFY and the process is conceptually "done". So maybe this is an atomic transition.

Reply:
Ok, I have added two new commits, one removing the `.outputClosing` state and one removing the `closeOutputPromise` — let me know what you think!